game-core/src/main/java/games/strategy/triplea/ai/tree/BattleTreeCalculator.java

```diff
@@ -0,0 +1,135 @@
+import games.strategy.triplea.delegate.Matches;
+import games.strategy.triplea.delegate.battle.MustFightBattle;
+import games.strategy.triplea.delegate.battle.UnitBattleComparator;
+import games.strategy.triplea.delegate.battle.casualty.CasualtyOrderOfLosses;
+import games.strategy.triplea.odds.calculator.AggregateResults;
+import games.strategy.triplea.odds.calculator.IBattleCalculator;
+import games.strategy.triplea.util.TuvUtils;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import org.triplea.java.collections.CollectionUtils;
+
+public class BattleTreeCalculator implements IBattleCalculator {
+
+  private GameData data;
+
+  public BattleTreeCalculator() {
+  }
+
+  public void setGameData(final GameData data) {
+    this.data = data;
+  }
+
+  @Override
+  public AggregateResults calculate(
+      final GamePlayer attacker,
+      final GamePlayer defender,
+      final Territory location,
+      final Collection<Unit> attackingUnits,
+      final Collection<Unit> defendingUnits,
+      final Collection<Unit> bombardingUnits,
+      final Collection<TerritoryEffect> territoryEffects,
+      final boolean retreatWhenOnlyAirLeft, // TODO: handle this flag
+      final int runCount
+  ) {
+    final long startTime = System.currentTimeMillis();
+    final int maxRounds =
+        location.isWater()
+            ? Properties.getSeaBattleRounds(data)
+            : Properties.getLandBattleRounds(data);
+    if (maxRounds > 0) {
+      BattleStep.MAX_ROUNDS = maxRounds;
+    }
+
+    // remove all of the non combatants
+    final MustFightBattle battle = new MustFightBattle(location, attacker, data, null);
+    final List<Unit> attackingUnitsCleaned = battle.removeNonCombatants(attackingUnits, true, true);
+    final List<Unit> defendingUnitsCleaned = battle.removeNonCombatants(defendingUnits, false, true);
+
+    final boolean amphibious = false;
+    final List<Unit> amphibiousLandAttackers = null;
+
+    final List<Unit> attackingOrderOfLoss =
+        CasualtyOrderOfLosses.sortUnitsForCasualtiesWithSupport(
+            CasualtyOrderOfLosses.Parameters.builder()
+                .targetsToPickFrom(attackingUnitsCleaned)
+                .player(attacker)
+                .enemyUnits(defendingUnitsCleaned)
+                .combatModifiers(
+                    UnitBattleComparator.CombatModifiers.builder()
+                        .territoryEffects(territoryEffects)
+                        .amphibious(amphibious)
+                        .defending(false)
+                        .build())
+                .amphibiousLandAttackers(
+                    amphibiousLandAttackers == null ? List.of() : amphibiousLandAttackers)
+                .battlesite(location)
+                .costs(TuvUtils.getCostsForTuv(attacker, data))
+                .data(data)
+                .build());
+
+    final List<Unit> defendingOrderOfLoss =
```

Comment:
Since the block of code is very similar to the one above, I think you should make a helper function for it with parameters for the things that are different.